#!/usr/bin/env python
# synergizer python API and example rpc calls
# copyright january 25 2012
# ross lazarus at gmail period com
# released under the LGPL
# exposes a pythonish API for synergizer descibed at http://llama.mshri.on.ca/synergizer/doc/


import json
import urllib2
import ast # safe conversion of potentially malicious code from incoming json

synergizerHost="http://llama.mshri.on.ca/cgi/synergizer/serv"

class synergizer:
    """
    synergizer API for converting gene names 
    eg print t.translate(authority="ncbi",domain="refseq_rna",species="Mus musculus", range="ensembl",ids=['NM_001001130','NM_001001144']) returns
    [['NM_001001130', 'ENSMUSG00000058331'], ['NM_001001144', 'ENSMUSG00000032485']]
    after getResponse is called, the results, id and error are available from the class instance so it can be reused as often as needed.
    Test exercises each API call with default parameters.
    """
    def __init__(self,host=synergizerHost,plist=[],pnames=["method","params","id"]):
        """ create the json from a list of parameters
        pnames is what Synergy wants as dict keys for the ordered params in plist
        """
        self.pnames = pnames
        self.plist = plist
        self.host = host
        
    
    def getResponse(self,plist=None):
        """
        idea from http://devinfee.com/blog/2010/11/08/make-json-requests-with-python/
        set up json call using this parameter list and existing pnames. parse returned string
        """
        assert len(plist) == len(self.pnames),'synergizer getResponse called without a %d element plist' % len(self.pnames)
        self.plist = plist # Overwrite before we call synergizer to update the instance
        self.jsndict = dict(zip(self.pnames,self.plist)) 
        self.jsn = json.dumps(self.jsndict)     
        req = urllib2.Request(self.host,self.jsn, {'content-type': 'application/json'})
        try:
            response_stream = urllib2.urlopen(req)
            response = response_stream.read()
        except:
            response = """{'id':'?','error':'interweb b0rken?',
                'result':'Cannot reach %s - is synergizer down or does the execution node have interweb access?' % self.host}"""
        self.response = response.replace('null','None')
        # NOTE: null is not a python built-in so ast.literal_eval fails so turn it into None
        rdict = ast.literal_eval('%s' % self.response) # safe eval !
        self.id = rdict.get('id','No id')
        self.result = rdict.get('result','No result')
        self.error = rdict.get('error','no error')
        self.rdict = rdict
        
    def getVers(self):
        """
        version()
        Returns current version of the Synergizer server as a string.
        """
        plist=["version",[],0]
        self.getResponse(plist=plist)
        return self.result

    def getAuth(self):
        """
        available_authorities()
        Returns a JSON array of strings corresponding
        to the currently available authorities.
        """
        plist = ["available_authorities",[],0]
        self.getResponse(plist=plist)
        return self.result


    def getAS(self,authority='ensembl'):
        """
        available_species( authority )
        This method takes as parameter a single string,
        representing an authority, and returns a JSON array of strings
        corresponding to the currently available species for the chosen authority.
        """
        plist=["available_species",[{"authority": authority}],0]
        self.getResponse(plist=plist)
        return self.result


    def getARange(self,authority='ncbi',species='Homo sapiens',domain="refseq_genomic"):
        """
        available_ranges( authority, species, domain )
        Takes as parameters three strings, representing an authority, a species,
        and a domain namespace, and returns a JSON array of strings corresponding to the
        currently available range namespaces for the chosen authority, species, and domain namespace.
        """
        plist=["available_ranges",[{"authority": authority,"species":species,"domain":domain}],0]
        self.getResponse(plist=plist)
        return self.result


    def getADomain(self,authority='ncbi',species='Homo sapiens'):
        """
        available_domains( authority, species )
        Takes as parameters two strings, representing an authority and a species, and
        returns a JSON array of strings corresponding to the currently available
        domain namespaces for the chosen authority and species.
        """
        plist=["available_domains", [{"authority": authority,"species": species}], 0]
        self.getResponse(plist=plist)
        return self.result

    def translate(self,authority='ncbi',species='Homo sapiens',domain='refseq_genomic',range="entrezgene",ids=['NM_001001130','NM_001001144']):
        """
        Main api call. ids should be a list.
        ids = ['NM_001001130','NM_001001144'] eg can be converted into ensembl ids with
        t.translate(authority="ncbi",domain="refseq_rna",species="Mus musculus", range="ensembl",ids=ids)
        """
        plist=["translate", [{"authority": authority,"species": species,"domain": domain,"range": range,"ids":ids}], 0]
        self.getResponse(plist=plist)
        return self.result
    
    def test(self):
        """
        original translate test case and examples for the other API calls
        """
        print 'version test result = ', self.getVers()
        print 'authority test result = ', self.getAuth()
        print 'domains test result = ', self.getADomain()
        print 'available species test result = ', self.getAS()
        print 'available range test result = ', self.getARange()
        # the translate example at http://llama.mshri.on.ca/synergizer/doc/
        # note that "null" is translated into None for pythonic sanity
        ids = ['NM_001001130','NM_001001144']
        print 'translate: ncbi: domain refseq_rna; Mus musculus, range:ensembl ids:%s' % ids, self.translate(authority="ncbi",
            domain="refseq_rna",species="Mus musculus", range="ensembl",ids=ids)
        ids = [ "snph", "chac1", "actn3", "maybe_a_typo","pja1", "prkdc", "RAD21L1", "Rorc", "kcnk16"]
        expected = [['snph', '9751'], ['chac1', '79094'], ['actn3', '89'], ['maybe_a_typo', None],
            ['pja1', '64219'], ['prkdc', '5591', '731751'], ['RAD21L1', '642636'], ['Rorc', '6097'], ['kcnk16', '83795']]
        tres = self.translate(authority="ensembl",domain="hgnc_symbol",range="entrezgene",ids=ids)
        testOK = (tres == expected)
        print 'Translate test result = ',testOK

    def generate_selects(self):
        """ need drop down lists in galaxy tool source - too slow to call the api
        """
        res = """
{'pombase': {'Schizosaccharomyces pombe': {'pombase': ['uniprotkb'], 'uniprotkb': ['pombase']}}, 'ncbi': {'Mycobacterium tuberculosis': {}, 'Canis familiaris': {}, 'Staphylococcus aureus': {}, 'Rattus norvegicus': {}, 'Xenopus laevis': {}, 'Danio rerio': {}, 'Macaca mulatta': {}, 'Zea mays': {'genbank_genomic_w_version': ['entrezgene', 'genbank_genomic', 'genbank_protein', 'genbank_protein_w_version', 'genbank_rna', 'genbank_rna_w_version', 'gi_genomic', 'gi_protein', 'gi_rna', 'maizegdb', 'refseq_protein', 'refseq_protein_w_version', 'refseq_rna', 'refseq_rna_w_version', 'uniprot'], 'gi_rna': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'gi_protein': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'genbank_rna_w_version': ['entrezgene', 'genbank_genomic', 'genbank_genomic_w_version', 'genbank_protein', 'genbank_protein_w_version', 'genbank_rna', 'gi_genomic', 'gi_protein', 'gi_rna', 'maizegdb', 'refseq_protein', 'refseq_protein_w_version', 'refseq_rna', 'refseq_rna_w_version', 'uniprot'], 'genbank_protein': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'refseq_protein': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'refseq_rna': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'genbank_genomic': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'genbank_protein_w_version': ['entrezgene', 'genbank_genomic', 'genbank_genomic_w_version', 'genbank_protein', 'genbank_rna', 'genbank_rna_w_version', 'gi_genomic', 'gi_protein', 'gi_rna', 'maizegdb', 'refseq_protein', 'refseq_protein_w_version', 'refseq_rna', 'refseq_rna_w_version', 'uniprot'], 'entrezgene': ['genbank_genomic', 'genbank_genomic_w_version', 'genbank_protein', 'genbank_protein_w_version', 'genbank_rna', 'genbank_rna_w_version', 'gi_genomic', 'gi_protein', 'gi_rna', 'maizegdb', 'refseq_protein', 'refseq_protein_w_version', 'refseq_rna', 'refseq_rna_w_version', 'uniprot'], 'uniprot': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'genbank_rna': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'refseq_rna_w_version': ['entrezgene', 'genbank_genomic', 'genbank_genomic_w_version', 'genbank_protein', 'genbank_protein_w_version', 'genbank_rna', 'genbank_rna_w_version', 'gi_genomic', 'gi_protein', 'gi_rna', 'maizegdb', 'refseq_protein', 'refseq_protein_w_version', 'refseq_rna', 'uniprot'], 'refseq_protein_w_version': ['entrezgene', 'genbank_genomic', 'genbank_genomic_w_version', 'genbank_protein', 'genbank_protein_w_version', 'genbank_rna', 'genbank_rna_w_version', 'gi_genomic', 'gi_protein', 'gi_rna', 'maizegdb', 'refseq_protein', 'refseq_rna', 'refseq_rna_w_version', 'uniprot'], 'gi_genomic': ['entrezgene', 'genbank_genomic_w_version', 'genbank_protein_w_version', 'genbank_rna_w_version', 'maizegdb', 'refseq_protein_w_version', 'refseq_rna_w_version'], 'maizegdb': ['entrezgene', 'genbank_genomic', 'genbank_genomic_w_version', 'genbank_protein', 'genbank_protein_w_version', 'genbank_rna', 'genbank_rna_w_version', 'gi_genomic', 'gi_protein', 'gi_rna', 'refseq_protein', 'refseq_protein_w_version', 'refseq_rna', 'refseq_rna_w_version', 'uniprot']}, 'Neisseria meningitidis': {}, 'Sus scrofa': {}, 'Triticum aestivum': {}, 'Yersinia pestis': {}, 'Ovis aries': {}, 'Geobacillus kaustophilus': {}, 'Bombyx mori': {}, 'Oncorhynchus mykiss': {}, 'Sinorhizobium meliloti': {}, 'Enterobacteria phage T4': {}, 'Drosophila melanogaster': {}, 'Dictyostelium discoideum': {}, 'Magnaporthe grisea': {}, 'Homo sapiens': {}, 'Equus caballus': {}, 'Caenorhabditis elegans': {}, 'Shigella flexneri': {}, 'Bos taurus': {}, 'Mus musculus': {}, 'Gallus gallus': {}, 'Trypanosoma brucei': {}, 'Pan troglodytes': {}, 'Strongylocentrotus purpuratus': {}, 'Glycine max': {}, 'Streptococcus pneumoniae': {}, 'Synechocystis 6803': {}, 'Thermus thermophilus': {}, 'Pseudomonas aeruginosa': {}, 'Thermotoga maritima': {}, 'Streptomyces coelicolor': {}, 'Aspergillus niger': {}, 'Solanum lycopersicum': {}, 'Felis catus': {}, 'Schizosaccharomyces pombe': {}, 'Bacillus subtilis': {}, 'Vibrio cholerae': {}, 'Escherichia coli': {}, 'Oryctolagus cuniculus': {}, 'Arabidopsis thaliana': {}, 'Salmonella enterica': {}, 'Chlamydomonas reinhardtii': {}, 'Saccharomyces cerevisiae': {}}, 'wormbase': {'Caenorhabditis elegans': {'intronerator': ['cds', 'cgc_name', 'eugenes', 'gene', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'protein_uniprot_ac': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'protein_uniprot_id': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'sequence_name': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'source_genbank', 'transcript'], 'public_name': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'sequence_name', 'source_genbank', 'transcript'], 'cgc_name': ['cds', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'cds': ['cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'source_genbank': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'transcript'], 'ndb_gi': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'protein': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'gene': ['cds', 'cgc_name', 'eugenes', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'transcript': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank'], 'protein_id': ['cds', 'cgc_name', 'eugenes', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript'], 'eugenes': ['cds', 'cgc_name', 'gene', 'intronerator', 'ndb_gi', 'protein', 'protein_id', 'protein_uniprot_ac', 'protein_uniprot_id', 'public_name', 'sequence_name', 'source_genbank', 'transcript']}}, 'cgd': {'Candida albicans': {'uniprotkb': ['cgd', 'entrezgene', 'refseq_rna'], 'entrezgene': ['cgd', 'refseq_rna', 'uniprotkb'], 'cgd': ['entrezgene', 'refseq_rna', 'uniprotkb'], 'refseq_rna': ['cgd', 'entrezgene', 'uniprotkb']}}, 'rgd': {'Rattus norvegicus': {'entrez_gene': ['ensembl_id', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'ensembl_id': ['entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'genbank_nucleotide': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'uniprot_id': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id'], 'symbol': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'unigene_id', 'uniprot_id'], 'genbank_protein': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'refseq_protein': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'refseq_rna': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'symbol', 'unigene_id', 'uniprot_id'], 'unigene_id': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna',
 'symbol', 'uniprot_id'], 'genbank': ['ensembl_id', 'entrez_gene', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'gene_rgd_id': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'refseq', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'refseq_genomic': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id'], 'refseq': ['ensembl_id', 'entrez_gene', 'genbank', 'genbank_nucleotide', 'genbank_protein', 'gene_rgd_id', 'refseq_genomic', 'refseq_protein', 'refseq_rna', 'symbol', 'unigene_id', 'uniprot_id']}}, 'ensembl': {'Erinaceus europaeus': {}, 'Xenopus tropicalis': {'uniprot_sptrembl': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'xenopus_jamboree': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession'], 'mirbase_accession': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_transcript_id', 'entrezgene', 'mirbase_id', 'unigene'], 'embl': ['affy_xtropicalis', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'mirbase_accession', 'mirbase_id', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'affy_xtropicalis': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'mirbase_accession', 'mirbase_id', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'rfam', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'refseq_peptide_predicted': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'protein_id': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'merops': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'ensembl_gene_id': ['affy_xtropicalis', 'embl', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'mirbase_accession', 'mirbase_id', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'rfam', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'refseq_peptide': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'entrezgene': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'merops', 'mirbase_accession', 'mirbase_id', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'rfam', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'uniprot_swissprot_accession': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'xenopus_jamboree'], 'uniprot_genename': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'ensembl_transcript_id': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'entrezgene', 'merops', 'mirbase_accession', 'mirbase_id', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'rfam', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'uniprot_swissprot': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'ensembl_peptide_id': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'rfam': ['affy_xtropicalis', 'ensembl_gene_id', 'ensembl_transcript_id', 'entrezgene', 'unigene'], 'mirbase_id': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_transcript_id', 'entrezgene', 'mirbase_accession', 'unigene'], 'pdb': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'unigene', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'unigene': ['affy_xtropicalis', 'embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'entrezgene', 'merops', 'mirbase_accession', 'mirbase_id', 'pdb', 'protein_id', 'refseq_peptide', 'refseq_peptide_predicted', 'rfam', 'uniprot_genename', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree']}, 'Canis familiaris': {}, 'Rattus norvegicus': {}, 'Dipodomys ordii': {}, 'Gorilla gorilla': {}, 'Vicugna pacos': {}, 'Danio rerio': {}, 'Macaca mulatta': {}, 'Otolemur garnettii': {}, 'Monodelphis domestica': {}, 'Myotis lucifugus': {}, 'Spermophilus tridecemlineatus': {}, 'Microcebus murinus': {}, 'Ochotona princeps': {}, 'Dasypus novemcinctus': {}, 'Drosophila melanogaster': {}, 'Procavia capensis': {}, 'Aedes aegypti': {}, 'Gasterosteus aculeatus': {}, 'Homo sapiens': {}, 'Loxodonta africana': {}, 'Equus caballus': {}, 'Caenorhabditis elegans': {}, 'Tarsius syrichta': {}, 'Bos taurus': {}, 'Mus musculus': {}, 'Gallus gallus': {}, 'Ciona savignyi': {}, 'Takifugu rubripes': {}, 'Echinops telfairi': {}, 'Oryzias latipes': {}, 'Pteropus vampyrus': {}, 'Tupaia belangeri': {}, 'Pan troglodytes': {}, 'Felis catus': {}, 'Tursiops truncatus': {}, 'Tetraodon nigroviridis': {}, 'Oryctolagus cuniculus': {}, 'Arabidopsis thaliana': {}, 'Ornithorhynchus anatinus': {}, 'Anopheles gambiae': {}, 'Saccharomyces cerevisiae': {}, 'Cavia porcellus': {}, 'Ciona intestinalis': {}, 'Sorex araneus': {}}, 'ecocyc': {'Escherichia coli': {'pfam': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'uniprot': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel'], 'refseq_protein': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecogene': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecoo157cyc': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecocyc_gene': ['asap', 'cgsc', 'echobase', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'asap': ['cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'modbase': ['asap', 'cgsc', 'echobase', 'ecocyc_gene',
 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'cgsc': ['asap', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'echobase': ['asap', 'cgsc', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecocyc_synonym': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ou-microarray': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecoliwiki': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecolihub': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecocyc_product': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'shigellacyc': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'swissmodel', 'uniprot'], 'swissmodel': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'uniprot'], 'sfle198214cyc': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'shigellacyc', 'swissmodel', 'uniprot'], 'pdb': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'ecol199310cyc': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'regulondb', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot'], 'regulondb': ['asap', 'cgsc', 'echobase', 'ecocyc_gene', 'ecocyc_product', 'ecocyc_synonym', 'ecogene', 'ecol199310cyc', 'ecolihub', 'ecoliwiki', 'ecoo157cyc', 'modbase', 'ou-microarray', 'pdb', 'pfam', 'refseq_protein', 'sfle198214cyc', 'shigellacyc', 'swissmodel', 'uniprot']}}, 'ensembl49': {'Erinaceus europaeus': {}, 'Xenopus tropicalis': {'xenopus_jamboree': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id'], 'uniprot_varsplic_id': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'xenopus_jamboree'], 'embl': ['ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'unified_uniprot': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'mirbase': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'uniprot_sptrembl': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'hgnc_id': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'unigene': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'ensembl_gene_id': ['embl', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'rfam': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'uniprot_swissprot_accession': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'unified_uniprot_accession': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'ensembl_transcript_id': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'uniprot_swissprot': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'ensembl_peptide_id': ['embl', 'ensembl_gene_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'hgnc_symbol': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'mirbase', 'pdb', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'pdb': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'protein', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree'], 'protein': ['embl', 'ensembl_gene_id', 'ensembl_peptide_id', 'ensembl_transcript_id', 'hgnc_id', 'hgnc_symbol', 'mirbase', 'pdb', 'rfam', 'unified_uniprot', 'unified_uniprot_accession', 'unigene', 'uniprot_sptrembl', 'uniprot_swissprot', 'uniprot_swissprot_accession', 'uniprot_varsplic_id', 'xenopus_jamboree']}, 'Canis familiaris': {}, 'Rattus norvegicus': {}, 'Echinops telfairi': {}, 'Danio rerio': {}, 'Macaca mulatta': {}, 'Otolemur garnettii': {}, 'Monodelphis domestica': {}, 'Myotis lucifugus': {}, 'Spermophilus tridecemlineatus': {}, 'Microcebus murinus': {}, 'Ochotona princeps': {}, 'Dasypus novemcinctus': {}, 'Drosophila melanogaster': {}, 'Pongo
 pygmaeus': {}, 'Aedes aegypti': {}, 'Gasterosteus aculeatus': {}, 'Homo sapiens': {}, 'Loxodonta africana': {}, 'Equus caballus': {}, 'Caenorhabditis elegans': {}, 'Bos taurus': {}, 'Mus musculus': {}, 'Gallus gallus': {}, 'Ciona savignyi': {}, 'Takifugu rubripes': {}, 'Tetraodon nigroviridis': {}, 'Oryzias latipes': {}, 'Tupaia belangeri': {}, 'Pan troglodytes': {}, 'Felis catus': {}, 'Oryctolagus cuniculus': {}, 'Ornithorhynchus anatinus': {}, 'Anopheles gambiae': {}, 'Saccharomyces cerevisiae': {}, 'Cavia porcellus': {}, 'Ciona intestinalis': {}, 'Sorex araneus': {}}, 'sgd': {'Saccharomyces cerevisiae': {'sgdid': ['sgdid_secondary', 'standard/reserved', 'systematic'], 'sgdid_secondary': ['sgdid', 'standard/reserved', 'systematic'], 'systematic': ['sgdid', 'sgdid_secondary', 'standard/reserved'], 'standard/reserved': ['sgdid', 'sgdid_secondary', 'systematic']}}}

        """
        authorities = self.getAuth()
        authorities.sort()
        dlist = [{} for x in authorities]
        sdict=dict(zip(authorities,dlist)) # make an authority keyed dict of dicts
        for a in authorities:
            aSpecies = self.getAS(authority=a)
            for s in aSpecies:
                aDomains = self.getADomain(authority=a,species=s)
                if aDomains:
                    sdict[a][s] = {} # for domain keys
                    print 'Getting ranges for ',aDomains
                    for d in aDomains:
                        aRange = self.getARange(authority=a,species=s,domain=d)
                        if aRange:
                            print 'Authority=',a,'Species=',s,'domain=',d,'aRange=',aRange
                            sdict[a][s][d] = aRange
        print sdict
            
        
if __name__ == "__main__":
        s = synergizer()
        s.generate_selects()
        s.test()
    

