<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
	"http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<title>Starting and Running Gmaj</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta http-equiv="Content-Style-Type" content="text/css">
<link rel="stylesheet" type="text/css" href="gmaj.css">
</head>
<body>
<p class=vvlarge>
<h2>Starting and Running Gmaj</h2>
<p class=vvlarge>
TABLE OF CONTENTS
<p class=small>
<ul class=notop>
<li><a href="#intro">Introduction</a>
<li><a href="#start">Starting Gmaj</a>
<li><a href="#memory">Memory Allocation</a>
<li><a href="#screen">Screen Layout</a>
<li><a href="#mouse">Mouse Controls</a>
<li><a href="#menu">Menus and Buttons</a>
<li><a href="#notes">Footnotes</a>
</ul>
<p class=vlarge>

<h3><a name="intro">Introduction</a></h3>
<p>
Gmaj can be run in two different modes: as an applet over the
world-wide web (for viewing data delivered from a server), or as
a stand-alone application (for viewing data stored on your own
computer).  These modes are mostly similar but have a few minor
differences, as noted below.
<p>

<h3><a name="start">Starting Gmaj</a></h3>
<p>
If you are using Gmaj in applet mode, it will be started for you
when you visit the applicable web page or submit a query to the
server.  If the Gmaj window does not appear automatically, just
click on the labeled button to view the indicated data.  Then
skip the rest of this section.
<p>
If you are using Gmaj in stand-alone mode, you need to start it
yourself.  The Java runtime environment does not have its own
GUI, so you generally need to run Gmaj from a command line (e.g.,
in the MS-DOS Prompt window on Windows 98).  The basic command to
type in looks like this:
<pre>
    [path1]java -jar [path2]gmaj.jar
</pre>
where <code>[path1]</code> is the location of your
<code>java</code> program file (perhaps <code>c:\windows\</code>
on Win98, or <code>/usr/bin/java/</code> on a Unix system), and
<code>[path2]</code> is the location where the
<code>gmaj.jar</code> file was installed.  Note that you can
leave off <code>[path1]</code> if you have set up your system
command path to include the location of the <code>java</code>
program.  Depending on how your system is set up, it may also be
possible to run the jar file directly by just typing its name or
double-clicking on it.
<p>
Since you haven't yet specified any data to display, Gmaj will
begin by presenting a dialog box to prompt you for the name of
your input file (see <a href="gmaj_input.html"
>Input Files for Gmaj</a>).  If the file is located in the
current directory you can just type its name, otherwise you'll
need to supply a complete path.  When you click the OK button,
a window will appear displaying the loaded data.
<p>
As an alternative to using the dialog box, you can specify the
input file (plus additional parameters) on the command line.
As of this writing, the full command syntax is:
<pre>
    [path1]java -jar [path2]gmaj.jar
        [-version] [-help] [-debug] [-urlpause &lt;millisec&gt;]
        [-initzoom &lt;refseqname&gt; &lt;start&gt; &lt;end&gt;]
        [-bundle &lt;zipfile&gt;] [&lt;paramfile&gt;|&lt;alignfile&gt;]
</pre>
<p>
This has been wrapped for easier readability, but should be
typed all on one line.  Arguments shown in square brackets
<code>[]</code> are optional, while a vertical bar <code>|</code>
indicates a choice between alternatives.  Angle brackets
<code>&lt;&gt;</code> signify meta-syntactic variables that
should be replaced with your names or numbers.  Don't type any
of the brackets or the bar.
<p>
These parameters do the following:
<dl>
<dt>	<code>-version</code>:
<dd>	Prints a message with information about Gmaj, including
	version, author, etc.; then exits.
<dt>	<code>-help</code>:
<dd>	Prints a brief help message with up-to-date syntax; then
	exits.
<dt>	<code>-debug</code>:
<dd>	Instructs Gmaj to print a few extra warning messages in
	your terminal window if certain problems occur.  Normally
	you won't need this, as it is mainly for development
	purposes.
<dt>	<code>-urlpause</code>:
<dd>	Specifies how many milliseconds the program should pause
	before retrieving each file from a URL, in order to avoid
	overloading the server.
<dt>	<code>-initzoom</code>:
<dd>	Specifies an initial zoom setting to be applied when the
	window opens.  You will still be able to invoke the Unzoom
	or Set Zoom features interactively to see the entire
	sequence range.  The <code>&lt;refseqname&gt;</code> must
	match one of the sequence names from the alignment file,
	and the endpoints do include the offset (if any) for that
	sequence from the parameters file.  To specify the reference
	sequence without a zoom region, use <code>-1</code> for both
	endpoints.
<dt>	<code>-bundle</code>:
<dd>	Specifies the name of a <code>.zip</code> or
	<code>.jar</code> file containing some or all of the data
	files.  This option is mostly used with Gmaj's applet mode
	to streamline the data download, but it is also supported
	in stand-alone mode.  It is described in
	<a href="gmaj_input.html">Input Files for Gmaj</a>.
<dt>	<code>&lt;paramfile&gt;</code>:
<dd>	This is the meta-data parameters file that lists the names
	of all the data files, plus a few related parameters such as
	display offsets and any intrinsic reference sequence.  For
	more information about the contents and format of this file,
	please see <a href="gmaj_input.html">Input Files for Gmaj</a>
	and <code><a href="sample.gmaj">sample.gmaj</a></code>.
<dt>	<code>&lt;alignfile&gt;</code>:
<dd>	If you don't want to use any annotations or other
	data-related options, you can simply specify the alignment
	file directly, instead of creating a parameters file.  This
	must be in <code>.maf</code> format; see
	<a href="gmaj_input.html">Input Files for Gmaj</a> for more
	details.
</dl>
<p>

<h3><a name="memory">Memory Allocation</a></h3>
<p>
If the dataset you want to view is large, you may need to give
Java more memory using the <code>-Xmx</code> switch (at least
with Sun's Java; this option may not be supported by all vendors).
<p>
For example, when using stand-alone mode the command line
<pre>
    [path1]java -Xmx1024m -jar [path2]gmaj.jar
</pre>
runs Gmaj with a heap memory allowance of ~1 gigabyte.  For the
applet, you can do this by opening the Java Plug-in Control Panel
on your computer, going to the Advanced tab, and entering
<code>-Xmx1024m</code> in the Java Runtime Parameters box (this
will affect all applets you run via the Java Plug-in, not just
Gmaj).
<p>

<h3><a name="screen">Screen Layout</a></h3>
<p>
The Gmaj window is divided into several sections.  Across the
top you will see a menu bar, and below that two lines containing
indicator boxes with information about the position of the mouse
pointer, the endpoints of the currently displayed zoom region,
and the location of the "mark" (<a href="#red">red</a> circle),
along with buttons for sliding the zoom region and selecting
alternative blocks at the marked position.  (The menus and
buttons will be discussed individually in the <a href="#menu"
>Menus and Buttons</a> section of this document.)
<p>
<i>Ruler:</i><br>
The first graphical panel is a horizontal ruler that displays
tick marks corresponding to positions in the currently selected
reference sequence.  These are intended to give you an immediate
general feel for the location and scale of the region being
displayed.  Precise locations can be determined via the position
indicator, which displays the exact coordinate of the mouse
pointer.
<p>
<i>Annotation links:</i><br>
Below the ruler is a panel that can display links to additional
information about various sequence regions.  Each annotation is
represented by a color-coded bar spanning the region's position
in the reference sequence.  (The bars' vertical positions are
not meaningful; they are only placed in rows for convenience, to
keep them from overlapping.)  Pointing to a particular bar will
cause the position indicator to display the <code>x</code>
coordinate of the pointer, and also the type and description of
that bar's annotation; otherwise only the <code>x</code>
coordinate will be shown.  In applet mode, clicking on a bar
will open a separate browser window to visit the corresponding
web site.  In stand-alone mode Gmaj is not working within a web
browser, so instead it displays the URL for you to visit manually
via copy-and-paste.  If you do not provide a links file, this
panel will not appear.
<p>
<i>Sequence features:</i><br>
The next two panels contain schematic diagrams of the known
exons and repeats in the current reference sequence, respectively
(if these files were provided).  Any additional features such as
CpG islands are included with the repeats.  Again, the position
indicator displays the <code>x</code> coordinate of the mouse
pointer, and also identifies any features at that position.
<p>
<i>Pips:</i><br>
The following panels display a number of pips (percent identity
plots) showing the pairwise alignments, projected from the
multiple alignment, of the reference sequence against each of
the other sequences.  A pip is similar to a dotplot, with the
horizontal <code>x</code>-axis representing the position in the
reference sequence, but the vertical <code>y</code>-axis
represents the percentage of matching nucleotides in each
gap-free segment of the pairwise alignment, instead of its
position in the second sequence.  Only the top half of each plot
is shown, since segments matching less than 50% are usually not
very interesting.  An additional feature of these panels is that
colored backgrounds, or "underlays", can be used to highlight
regions of interest (if you provide files with this information).
Vertical blue bars at the edges of the pips represent the
boundaries of the current zoom region, whose endpoints are
displayed in the zoom indicator.  The position indicator displays
the horizontal coordinate and vertical percentage position of the
mouse pointer, along with a list of <a href="#block">block
numbers</a> covering that location.  It will also display labels
for the colored regions if these are included in the underlay
files.
<p>
<i>Text view:</i><br>
The bottom panel displays a nucleotide-level view of a single
selected alignment block.  (Initially it is empty, since you
haven't selected anything yet.)  The top row of this display
shows the current reference sequence, while the rows for the
other sequences show a dot "<code>.</code>" wherever they match
the reference sequence, and only explicitly list the nucleotides
that don't match.  (This matching is case-insensitive to deal
with soft masking, but non-nucleotide characters such as
<code>X</code> or <code>N</code> never match anything, even
themselves.)  All of the sequences will likely have had gaps
"<code>-</code>" inserted by the alignment program.  Note that
most of the blocks will be much too long to fit across this
window, so a scrollbar is provided; the relative size of the
scrollbar's slider indicates what fraction of the alignment is
shown in the window.  Shaded "highlights" (analogous to the pip
underlays) can also be specified for each sequence; otherwise
Gmaj will provide default highlights based on the exons files
(if any).  Whenever the mouse pointer is in this bottom panel,
the position indicator displays its location in the format
"<code>n:x</code>", where <code>n</code> is the column position
in the aligned block (starting with 0), and <code>x</code> is the
sequence position in the individual row (starting with 1).  Note
that <code>x</code> does not include the gaps, but <code>n</code>
does.  Labels for any highlights at that position are also
displayed.
<p>
With the exception of the text view, all of these panels use the
same horizontal coordinate scale (i.e., position in the current
reference sequence), and they are always kept vertically aligned
so they can be compared easily.  Note that the partition between
the graphical panels and the text view is draggable, so you can
adjust the relative amount of space they occupy.  Also,
individual panels can be hidden if desired, using the Options -
Show dialog (see <a href="#menu">Menus and Buttons</a>).
<p>
<!--
<i>Dotplot:</i><br>
The large middle panel displays a dotplot view of the alignments,
with the reference sequence along the horizontal
<code>x</code>-axis and the secondary sequence along the vertical
<code>y</code>-axis.  If the second sequence contains multiple
contigs, they will appear as separate horizontal bands across the
plot, each with its own <code>y</code>-axis coordinate system.
Whenever the mouse pointer is in this panel, the position
indicator displays its location in the format "<code>x,y</code>",
where <code>x</code> is the position in the horizontal sequence
and <code>y</code> is the position in the vertical sequence.  If
there are multiple contigs, then the contig name will be
displayed as well (actually only the first word is displayed, to
prevent long names from crowding out the other information).
<p>
-->

<h3><a name="mouse">Mouse Controls</a></h3>
<p>
You can select a particular alignment block by clicking on one
of its segments in any of the pips with the left mouse button.
(Actually you don't have to click exactly on it, because
Gmaj will automatically jump to the nearest point if you miss.)
The spot will be marked with a small <a href="#red">red</a>
circle, and the entire alignment block containing the mark will
change color from black to <a href="#red">red</a> in all of the
pips (each block typically spans several gap-free segments).
Also, the corresponding text view for that block will appear in
the bottom panel with the selected position highlighted.  Lastly,
the mark indicator will be filled in with information about the
marked block and position, and a row of buttons will appear next
to it showing the <a href="#block">block numbers</a> covering
the marked location.  These buttons allow convenient selection
of a different block at the same position (see <a href="#menu"
>Menus and Buttons</a>).  Note that there is only one mark at a
time in each window, so the previous one, if any, will be
unmarked.
<p>
In a similar fashion, clicking the left mouse button in the
text view will move the mark (both the highlight and the
<a href="#red">red</a> circle) to that position.  However, gap
positions cannot be selected in this manner because they do not
correspond to pip segments; if you click in a gap, the nearest
gap-free position is selected instead.  Also, if you click on a
position in the reference sequence (which has no corresponding
pip), the mark will move to the new column but will remain in
the same pip as before.
<p>
You can "zoom in" on a particular region by dragging out a
rectangle with the left mouse button in any of the white panels
(ruler, links, exons, repeats, or pip).  All of these panels
will always zoom together, to keep them lined up.  This can be
repeated until the maximum resolution is reached; after that
Gmaj will display an error message.  Additional zoom features
are available via the Zoom menu and arrow buttons (see
<a href="#menu">Menus and Buttons</a>).  Note that selecting
a new region will cause any entries in your zoom history that
are forward of the current point to be discarded (similar to
a web browser).
<p>
Holding down the right mouse button over any of the white
panels adds crosshairs at the mouse pointer's location, which
is convenient for determining whether two regions really line
up.  If you have a one-button mouse, you can achieve the same
effect by applying the <code>Shift</code> key when initially
pressing the mouse button.
<p>

<h3><a name="menu">Menus and Buttons</a></h3>
<p>
<dl>
<dt>File - Open:
<dd>
Loads a new set of data files into Gmaj, replacing the currently
displayed data.  A dialog box is presented for you to enter the
new input file.  (See discussion under <a href="#start"
>Starting Gmaj</a>, above.)  This menu item does not appear in
applet mode, because the user is unlikely to know the locations
of other data files on the server; instead the webmaster should
set up separate access for each dataset.
<p>
<dt>File - Export:
<dd>
Saves the currently selected block (or all blocks overlapping
the current zoom region) to disk as FastA sequences.  A separate
file is created for each sequence name (i.e. species or contig).
Options include clipping the sequences to the zoom region
boundary, and restoring those that align in reverse complement
to their original orientation (which will also swap the order of
the endpoint coordinates in the FastA header).  Currently this
feature is not available in applet mode, because security
restrictions make it very awkward to save files on the local
computer.
<p>
<dt>File - Close:
<dd>
Closes the current window.
<p>
<dt>File - Exit:
<dd>
Exits from Gmaj.  In stand-alone mode, also exits from Java.
<p>
<dt>Options:
<dd>
This menu controls some of the aesthetic aspects of Gmaj.  You
can choose between two sizes of fonts, which will also affect
some other visibility settings, such as the radius and thickness
of the mark circle, the blackness of the ruler numbers, and the
height of the pips.  You can also choose to make the mark circle
and the selected block's pip segments always red, instead of
changing with the background color (this is easier to explain in
a figure legend, but makes them invisible against red underlays).
Lastly, the Show item opens a dialog where you can choose which
individual panels to display or hide (currently it does not
affect which text rows are displayed, however).  The choices on
this menu affect all of the windows, not just the current one.
<p>
<dt>RefSequence:
<dd>
This menu allows you to select a different reference sequence
(unless the parameters file indicates that the alignment has a
fixed, intrinsic reference sequence).  A new window will open,
showing the same data from the perspective of the sequence you
chose.  The mark (if any) will be copied to the new window as
closely as possible, but the zoom will not.  Thereafter, the
windows will operate independently.  You can have one window
open for each sequence in the data.
<p>
<dt>Zoom - Back:
<dd>
Moves backward in your zoom history for this window, returning
to previous regions.  Does not affect the mark.
<p>
<dt>Zoom - Forward:
<dd>
Moves forward in your zoom history for this window.  Does not
affect the mark.
<p>
<dt>Zoom - Unzoom:
<dd>
Sets the zoom region to the widest, unzoomed view, i.e., the
full range for this reference sequence as specified in the
<code>.maf</code> file.  Has the same effect as entering the
"valid range" endpoints in Set Zoom.  Does not affect the mark.
<p>
<dt>Zoom - Set Zoom:
<dd>
Presents a dialog box that allows you to enter arbitrary zoom
endpoints (within the valid range for this reference sequence).
The new region, if different from the current one, is added to
your zoom history for this window.  Any regions forward of the
current point in your history are discarded (similar to a web
browser).  Does not affect the mark.
<p>
<dt>Help - About:
<dd>
Displays a message window with information about Gmaj, including
version, author, etc.  Also reports the version of Java you are
currently using.
<p>
<dt>Help - Manual:
<dd>
In applet mode, opens a new browser window to view this help
page.  In stand-alone mode Gmaj is not working within a web
browser, so instead it displays the URL for you to visit manually
via copy-and-paste.
<p>
<dt>Help - Keys:
<dd>
Displays a message window listing Gmaj's keyboard shortcuts.  No
<code>Alt</code> key is needed, since Gmaj doesn't use the
keyboard for much else.
<p>
<dt>Arrow Buttons:
<dd>
These two buttons are located to the right of the zoom indicator.
Clicking on one of them will slide the zoom region to the left or
right, respectively, by half of its width.  The new region is
added to your zoom history like any other zoom selection, so the
Zoom - Back command will return to where you were as usual.
<p>
<dt>Block Buttons:
<dd>
When a block is selected by clicking on a pip, a row of buttons
will appear to the right of the mark indicator showing all of the
<a href="#block">block numbers</a> covering the marked position.
(If there is not enough room for all of the buttons, a scrollbar
will appear; also the partition between this panel and the mark
indicator is draggable.)  Clicking on one of the buttons causes
the mark to move to that block (in the same pip), and the pip
segment colors, text alignment, and mark indicator will be
updated accordingly.  The new marked position (red circle) will
be as close as possible to the same coordinate in the reference
sequence, but it may have to move slightly to avoid gaps.  This
makes it theoretically possible, though rare, that the resulting
block list (and therefore the row of buttons) may change.
</dl>
<p>

<!-- <hr align=left noshade size=1 width="20%" color=black> -->
<h3><a name="notes">Footnotes</a></h3>
<p>
<a name="red"></a>
[1]&nbsp;
The circular mark and the selected block's pip segments are red
when the background is white, but by default are displayed in
different colors against other backgrounds to ensure good
contrast.  A setting on the Options menu can make them always
red if desired.  The selected pip segments are drawn last, so
they will not be obscured by other blocks.
<p>
<a name="block"></a>
[2]&nbsp;
An alignment block is considered to cover a pip position if it
contains rows for both of the pip's sequences and the position
falls within the endpoints of the reference sequence's row; there
are no "holes" due to gaps.  Blocks in the <code>.maf</code> file
are numbered consecutively, starting with 0.
<p>

<p class=vvlarge>
<hr>
<i>Cathy Riemer, June 2006</i>

<p class=scrollspace>
</body>
</html>
