<html>
<head><title>Galaxy data formats</title>
</head>
<body>
<h2>Galaxy data formats</h2>
<p>
For problems with missing queries in the tool selection boxes the most common
reason is the tool only lists history items with data formats compatible with the 
tool.  Some formats are subsets of others and Galaxy should also list those
with compatible subformats as well.  If the query is not showing up still and you 
believe it is in the correct format you can click on the pencil icon and 
manually change the format.  This will not edit the file just change the
metadata for the file.  Some cases you will need to actually change the 
file format.  For example, if the file is space delimited and a
tabular file is required; then the "Convert delimiters to TAB" tool under "Text 
Manipulation" can be used to reformat the file.  
<p>
Some of the most commonly used formats are very similar.  Start with the 
basic tabular file.  It has few requirements other than 1 or more columns of
data separated by tabs.  Next is intervals which are tabular but they have the
added requirement that 3 of the columns must be the chromosome, start point,
and end point.  There is optionally a strand and header labelling the 
columns.  Next is BED or GFF, which are also tabular and intervals, but with more
restrictions.  BED can vary between 3 and 12 columns, with each being
precisely defined.  Here
the order of the columns also matters, and only the end columns can be skipped.
Some groups of the columns have to be all there or all left off.  GFF is
similar in setup but with all 9 columns required and different definitions.  
See more detailed descriptions below.
<hr>
<H2>Formats</H2>
<ul>
<li><a href="#ab1">Ab1</a>
<li><a href="#axt">AXT</a>
<li><a href="#bam">BAM</a>
<li><a href="#binseq">Binseq.zip</a>
<li><a href="#bed">BED</a>
<li><a href="#bedgraph">BedGraph</a>
<li><a href="#fasta">FASTA</a>
<li><a href="#fastqsolexa">FastqSolexa</a>
<li><a href="#fped">fped</a>
<li><a href="#gff">GFF</a>
<li><a href="#gff3">GFF3</a>
<li><a href="#gtf">GTF</a>
<li><a href="#html">Html</a>
<li><a href="#interval">Interval</a>
<li><a href="#lav">LAV</a>
<li><a href="#lped">Lped</a>
<li><a href="#maf">MAF</a>
<li><a href="#pbed">pbed</a>
<li><a href="#psl">PSL</a>
<li><a href="#scf">Scf</a>
<li><a href="#sff">Sff</a>
<li><a href="#table">Table</a>
<li><a href="#tab">Tabular</a>
<li><a href="#txtseqzip">Txtseq.zip</a>
<li><a href="#wig">Wiggle custom track</a>
<li><a href="#text">Other text type</a>
</ul>
<br>
<hr>
<strong>Ab1</strong>
<a name="ab1"/>
<p>
A binary sequence file in 'ab1' format with a '.ab1' file extension.  You must manually select this 'File Format' when uploading the file.

<hr/>

<strong>AXT</strong>
<a name="axt"/>
<p>
blastz pairwise alignment format.  Each alignment block in an axt file contains three lines: a summary line and 2 sequence lines.  Blocks are separated from one another by blank lines.  The summary line contains chromosomal position and size information about the alignment. It consists of 9 required fields.
<a href="http://main.genome-browser.bx.psu.edu/goldenPath/help/axt.html">Click here</a> for more information about axt format.
<dl><dt>Can be converted to:
<dd><ul>
<li>FASTA<br/>
Convert Formats&rarr;AXT to FASTA
<li>LAV<br/>
Convert Formats&rarr;AXT to LAV
</ul></dl>
<hr/>

<strong>Bam</strong>
<a name="bam"/>
<p>
A binary file compressed in the BGZF format with a '.bam' file extension.
<a href="http://samtools.sourceforge.net/SAM1.pdf">SAM</a>
format is the human readable text version of these files.
<dl><dt>Can be converted to:
<dd><ul>
<li>pileup<br/>
NGS: SAM Tools&rarr;Generate pileup<br/>
<li>interval<br/>
First you have to go to pileup as above then
NGS: SAM Tools&rarr;Pileup-to-Interval
</ul></dl>
<hr/>

<strong>Binseq.zip</strong>
<a name="binseq"/>
<p>
A zipped archive consisting of binary sequence files in either 'ab1' or 'scf' format.  All files in this archive must have the same file extension which is one of '.ab1' or '.scf'.  You must manually select this 'File Format' when uploading the file.

<hr/>

<strong>BED</strong>
<a name="bed"/>
<p>
<ul>
<li> also tabular
<li> also interval
</ul>
This describes a genomic interval, but has strict field specifications for use
in browsers.
<a href="http://main.genome-browser.bx.psu.edu/goldenPath/help/hgTracksHelp.html#BED">Click here</a> for field specifications.<br/>
Example:
<pre>
chr22 1000 5000 cloneA 960 + 1000 5000 0 2 567,488, 0,3512
chr22 2000 6000 cloneB 900 - 2000 6000 0 2 433,399, 0,3601
</pre>
<dl><dt>Can be converted to:
<dd><ul>
<li>GFF<br/>
Convert Formats&rarr;BED-to-GFF
</ul></dl>
<hr/>

<strong>BedGraph</strong>
<a name="bedgraph"/>
<p>
<ul>
<li> also tabular
<li> also interval
<li> also BED
</ul>
<a href="http://main.genome-browser.bx.psu.edu/goldenPath/help/bedgraph.html">BedGraph</a>
is a BED file with the name column being a float value that is displayed 
as a Wiggle in tracks.  Unlike wiggles this score can be retrieved in its 
exact value after being loaded as a track.
<hr/>

<strong>Fasta</strong>
<a name="fasta"/>
<p>
A sequence in 
<a href="http://www.ncbi.nlm.nih.gov/blast/fasta.shtml">FASTA format</a>
consists of a single-line description, followed by lines of sequence data.  The first character of the description line is a greater-than ("&gt;") symbol in the first column.  All lines should be shorter than 80 characters::
<pre>
>sequence1
atgcgtttgcgtgc
gtcggtttcgttgc
>sequence2
tttcgtgcgtatag
tggcgcggtga
</pre>
<dl><dt>Can be converted to:
<dd><ul>
<li>tabular<br/>
Convert Formats&rarr;FASTA-to-Tabular
</ul></dl>
<hr/>

<strong>FastqSolexa</strong>
<a name="fastqsolexa"/>
<p>
<a href="http://maq.sourceforge.net/fastq.shtml">FastqSolexa</a>
is the Illumina (Solexa) variant of the Fastq format, which stores sequences and quality scores in a single file
<pre>
@seq1  
GACAGCTTGGTTTTTAGTGAGTTGTTCCTTTCTTT  
+seq1  
hhhhhhhhhhhhhhhhhhhhhhhhhhPW@hhhhhh  
@seq2  
GCAATGACGGCAGCAATAAACTCAACAGGTGCTGG  
+seq2  
hhhhhhhhhhhhhhYhhahhhhWhAhFhSIJGChO

Or 

@seq1
GAATTGATCAGGACATAGGACAACTGTAGGCACCAT
+seq1
40 40 40 40 35 40 40 40 25 40 40 26 40 9 33 11 40 35 17 40 40 33 40 7 9 15 3 22 15 30 11 17 9 4 9 4
@seq2
GAGTTCTCGTCGCCTGTAGGCACCATCAATCGTATG
+seq2
40 15 40 17 6 36 40 40 40 25 40 9 35 33 40 14 14 18 15 17 19 28 31 4 24 18 27 14 15 18 2 8 12 8 11 9
</pre>
<dl><dt>Can be converted to:
<dd><ul>
<li>FASTA<br/>
Convert Formats&rarr;FASTQ to FASTA    
</ul></dl>
<hr/>

<strong>fped</strong>
<a name="fped"/>
<p>
Also known as the FBAT format, for use in the <a href="http://biosun1.harvard.edu/~fbat/fbat.htm">FBAT</a> program.  
It consists of a pedigree file and an phenotype file.
<hr/>

<strong>Gff</strong>
<a name="gff"/>
<p>
<ul>
<li> also tabular
<li> also interval
</ul>
GFF lines have 
<a href="http://main.genome-browser.bx.psu.edu/FAQ/FAQformat#format3">nine required fields</a> that must be tab-separated.
<dl><dt>Can be converted to:
<dd><ul>
<li>BED<br/>
Convert Formats&rarr;GFF-to-BED
</ul></dl>
<hr/>

<strong>GFF3</strong>
<a name="gff3"/>
<p>
<ul>
<li> also tabular
<li> also interval
</ul>
The 
<a href="http://www.sequenceontology.org/gff3.shtml">GFF3</a>
format addresses the most common extensions to GFF, while preserving backward compatibility with previous formats.
<hr/>

<strong>GTF</strong>
<a name="gtf"/>
<p>
<ul>
<li> also tabular
<li> also interval
</ul>
<a href="http://main.genome-browser.bx.psu.edu/FAQ/FAQformat#format4">GTF</a>
is a format for describing genes and other features associated with DNA, RNA and Protein sequences.
<dl><dt>Can be converted to:
<dd><ul>
<li>BED graph<br/>
Convert Formats&rarr;GTF-to-BEDGraph
</ul></dl>
<hr/>

<strong>Html</strong>
<a name="html"/>
<p>
This format is a html web page.  Click the eye icon to view the dataset in 
your browser.
<hr/>

<strong>Interval (Genomic Intervals)</strong>
<a name="interval">
<p>
<ul>
<li> also tabular
</ul>
Required fields
<ul>
<li>CHROM - The name of the chromosome (e.g. chr3, chrY, chr2_random) or contig (e.g. ctgY1).
<li>START - The starting position of the feature in the chromosome or contig. The first base in a chromosome is numbered 0.
<li>END - The ending position of the feature in the chromosome or contig. The chromEnd base is not included in the display of the feature. For example, the first 100 bases of a chromosome are defined as chromStart=0, chromEnd=100, and span the bases numbered 0-99.
</ul>
Optional
<ul>
<li>STRAND - Defines the strand - either '+' or '-'.
<li>Headers
</ul>
Example:
<pre>
    #CHROM START END   STRAND NAME COMMENT
    chr1   10    100   +      exon myExon
    chrX   1000  10050 -      gene myGene
</pre>
<dl><dt>Can be converted to:
<dd><ul>
<li>BED<br/>
The exact changes needed and tools to run can vary with what fields are in
the interval file and what size BED you are converting to.  In general you
will likely use Text Manipulation&rarr;Compute, Cut or Merge Columns.
</ul></dl>
<hr/>

<strong>LAV</strong>
<a name="lav"/>
<p>
<a href="http://www.bx.psu.edu/miller_lab/dist/lav_format.html">LAV</a>
is the primary output format for BLASTZ.  The first line of a .lav file begins with #:lav..
<dl><dt>Can be converted to:
<dd><ul>
<li>BED<br/>
Convert Formats&rarr;LAV to BED
</ul></dl>
<hr/>

<strong>Lped</strong>
<a name="lped"/>
<p>
This is the linkage pedigree format (separate map and ped files).  
These files together describe SNPs, the map file has the position and an
identifier for the SNP and the pedigree file has the alleles.
To upload this format into Galaxy do not use auto-detect for the file
format, instead select lped.  You will then be given two sections for uploading
files, one for the pedigree file and one for the map file.
For more information see <a href="http://www.broadinstitute.org/science/programs/medical-and-population-genetics/haploview/input-file-formats-0">linkage pedigree</a>
or <a href="http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#map">map</a>
or <a href="http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#ped">ped</a>.
<dl><dt>Can be converted to:
<dd><ul>
<li>pbed<br/>Automatic
<li>fped<br/>Automatic
</ul></dl>
<hr/>

<strong>MAF</strong>
<a name="maf"/>
<p>
TBA and multiz multiple alignment format.  The first line of a .maf file begins with ##maf. This word is followed by white-space-separated "variable=value pairs". There should be no white space surrounding the "=".
<a href="http://main.genome-browser.bx.psu.edu/FAQ/FAQformat#format5">Click here</a> for more about MAF format.
<dl><dt>Can be converted to:
<dd><ul>
<li>BED<br/>
Convert Formats&rarr;Maf to BED
<li>Interval<br/>
Convert Formats&rarr;Maf to Interval
<li>FASTA<br/>
Convert Formats&rarr;Maf to FASTA
</ul></dl>
<hr/>

<strong>pbed</strong>
<a name="pbed"/>
<p>
This is the binary version of the lped file format.
<dl><dt>Can be converted to:
<dd><ul>
<li>lped<br/>Automatic
</ul></dl>
<hr/>

<strong>PSL</strong>
<a name="psl"/>
<p>
<a href="http://main.genome-browser.bx.psu.edu/FAQ/FAQformat#format2">PSL</a>
format is for alignments, it is returned by 
<a href="http://genome.ucsc.edu/cgi-bin/hgBlat?command=start">BLAT</a>.
It does not include any sequence.
<hr/>

<strong>Scf</strong>
<a name="scf"/>
<p>
A binary sequence file in 'scf' format with a '.scf' file extension.  You must manually select this 'File Format' when uploading the file.
<a href="http://staden.sourceforge.net/manual/formats_unix_2.html">Click here</a>
for more information.
<hr/>

<strong>Sff</strong>
<a name="sff"/>
<p>
A binary file in 'Standard Flowgram Format' with a '.sff' file extension.
<dl><dt>Can be converted to:
<dd><ul>
<li>FASTA<br/>
Convert Formats&rarr;SFF converter
<li>FASTQ<br/>
Convert Formats&rarr;SFF converter
</ul></dl>
<hr/>

<strong>Table</strong>
<a name="table"/>
<p>
Text delimited into columns by something other than a tab.
<hr/>

<strong>Tabular (tab delimited)</strong>
<a name="tab"/>
<p>
Any data in tab delimited format (tabular)
<dl><dt>Can be converted to:
<dd><ul>
<li>FASTA<br/>
Convert Formats&rarr;Tabular-to-FASTA<br/>
Tabular file must have a title and sequence column.
<li>interval<br/>
If the tabular file has the chromosome, or is all on one chromosome, and
a position you can create an interval file.  If all one chromosome use
Text Manipulation&rarr;Add column to add the chromosome.  If the given position
is a 1 based position use Text Manipulation&rarr;Compute and the position
column minus 1 to get the start.  Otherwise do plus 1 to get the end.
</ul></dl>
<hr/>

<strong>Txtseq.zip</strong>
<a name="txtseqzip"/>
<p>
A zipped archive consisting of flat text sequence files.  All files in this archive must have the same file extension of '.txt'.  You must manually select this 'File Format' when uploading the file.

<hr/>

<strong>Wiggle custom track</strong>
<a name="wig"/>
<p>
The wiggle format is line-oriented.  Wiggle data is preceded by a track definition line, which gives the type of wiggle.  There are 3 different types, each
with their uses.  
<a href="http://main.genome-browser.bx.psu.edu/goldenPath/help/wiggle.html">More information here.</a>
<dl><dt>Can be converted to:
<dd><ul>
<li>interval<br/>
Convert Formats&rarr;Wiggle-to-Interval<br/>
As a second step this could be converted to BED 3 or 4 by removing columns.
Text Manipulation&rarr;Cut columns from a table
</ul></dl>
<hr/>

<strong>Other text type</strong>
<a name="text"/>
<p>
Any text file
<dl><dt>Can be converted to:
<dd><ul>
<li>tabular<br/>
If this is space or some other delimiter separated fields it can be 
converted to tabular.  Text Manipulations&rarr;Convert delimiters to TAB
</ul></dl>
<!-- blank lines so internal links will jump farther to end -->
<br/><br/><br/><br/><br/><br/><br/><br/><br/><br/><br/><br/>
</body>
</html>
